retranmission.go

```diff
@@ -0,0 +1,203 @@
+// that peers are in consistent state in terms of message communication.
+type retransmitter struct {
+	// storage is a message storage which is needed to store the messages
+	// which weren't acked and restore them after node restarts, in order to
+	// send them to other side again.
+	storage channeldb.MessageStore
+
+	// codeToIndex map is used to locate which messages can be deleted from
+	// the message storage in response to a retrieved ACK message. The
+	// mapping for items is code -> {index #1 , ... , index #n}. So when
+	// receiving a new message, we check for the existence of the message
+	// code in this index bucket, then delete all the messages from the
+	// top-level bucket that are returned.
+	codeToIndex map[lnwire.MessageCode][]uint64
+
+	// messagesToRetransmit list of messages that should be retransmitted
+	// to other side.
+	messagesToRetransmit []lnwire.Message
+
+	mutex sync.RWMutex
+}
+
+// newRetransmitter creates new instance of retransmitter.
+func newRetransmitter(storage channeldb.MessageStore) (*retransmitter, error) {
+
+	// Retrieve the messages from the message storage with their
+	// associated indexes.
+	indexes, messages, err := storage.Get()
+	if err != channeldb.ErrPeerMessagesNotFound && err != nil {
+		return nil, err
+	}
+
+	// Initialize map of code to index map, so than later we can retrieve
+	// indexes of messages that should be removed.
+	codeToIndex := make(map[lnwire.MessageCode][]uint64)
+	for i, message := range messages {
+		codeToIndex[message.Command()] = append(
+			codeToIndex[message.Command()],
+			indexes[i],
+		)
+	}
+
+	return &retransmitter{
+		storage:              storage,
+		codeToIndex:          codeToIndex,
+		messagesToRetransmit: messages,
+	}, nil
+}
+
+// Register adds message that should be acknowledged in the message storage.
+func (rt *retransmitter) Register(msg lnwire.Message) error {
+	switch msg.Command() {
+	// messages without acknowledgment
+	case lnwire.CmdCloseComplete,
+		lnwire.CmdChannelUpdateAnnouncement,
+		lnwire.CmdChannelAnnouncement,
+		lnwire.CmdNodeAnnouncement,
+		lnwire.CmdPing,
+		lnwire.CmdPong,
+		lnwire.CmdErrorGeneric,
+		lnwire.CmdInit:
+		return nil
+	default:
+		// Adds message to storage and returns the message index which
+		// have been associated with this message within the storage.
+		index, err := rt.storage.Add(msg)
+		if err != nil {
+			return err
+		}
+
+		// Associate the message index within the message storage
+		// with message code in order to remove messages by index later.
+		rt.mutex.Lock()
+		rt.codeToIndex[msg.Command()] = append(
+			rt.codeToIndex[msg.Command()], index,
+		)
+		rt.mutex.Unlock()
+
+		return nil
+	}
+}
+
+// Ack encapsulates the specification logic about which messages should be
+// acknowledged by receiving this one.
+func (rt *retransmitter) Ack(msg lnwire.Message) error {
+	switch msg.Command() {
+
+	case lnwire.CmdSingleFundingResponse:
```

Comment:
For now, all funding message should be omitted from retransmission other than the `FundingLocked` message. While testing locally I just hit a bug that causes the funding manager to deadlock if `lnd` is restarted mid a >1 conf required channel opening. 

Atm, the spec is incorrect. No funding messages until the point in which either side is committed to a funding transaction should be retransmitted at all.